Restrict vscode-file requests to main window

mjbvz opened the pull request:
This change tries to limit use of the `vscode-file` protocol to the main window.

bpasero commented on src/vs/platform/windows/electron-main/window.ts:
```diff
@@ -428,6 +428,8 @@ export class CodeWindow extends Disposable implements ICodeWindow {
 			this.dispose();
 		});
 
+		// Filter web requests
```
Is this just a headline or is there some code missing?

bpasero commented on src/vs/platform/windows/electron-main/window.ts:
```diff
@@ -441,16 +443,45 @@ export class CodeWindow extends Disposable implements ICodeWindow {
+
+			if (uri.scheme === Schemas.vscodeFileResource) {
+				if (!isAllowedVsCodeFileRequest(uri, details)) {
+					console.log('Blocked vscode-file request', details.url);
```
Should this use our logger rather with `logService.error` similar to our protocol handling: https://github.com/microsoft/vscode/blob/ee1655a82ebdfd38bf8792088a6602c69f7bbd94/src/vs/platform/protocol/electron-main/protocolMainService.ts#L77?

bpasero commented on src/vs/platform/windows/electron-main/window.ts:
```diff
@@ -441,16 +443,45 @@ export class CodeWindow extends Disposable implements ICodeWindow {
+		const workbenchRootUrl = FileAccess.asBrowserUri('vs/', require).toString(true);
+		const isAllowedVsCodeFileRequest = (uri: URI, details: Electron.OnBeforeRequestListenerDetails) => {
+			// Allow vscode-file requests from main window
+			if (details.frame?.url.startsWith('vscode-file://vscode-app/')) {
```
Maybe @deepak1556 has an idea if this check can be improved instead of probing on the URL.

deepak1556 replied:
A frame is tracked in chromium by mapping the id of the process it belongs and a unique id within that process since there can be multiple frames in the same process. So always checking [frame.processId](https://github.com/electron/electron/blob/main/docs/api/web-frame-main.md#frameprocessid-readonly) and [frame.routingId](https://github.com/electron/electron/blob/main/docs/api/web-frame-main.md#frameroutingid-readonly) will be right way to identify a frame.

In this case, we want to compare the above properties of incoming frame with [this._win.webContents.mainFrame](https://github.com/electron/electron/blob/main/docs/api/web-contents.md#contentsmainframe-readonly) properties to allow requests for top level frame.

If we want to allow any frames from the workbench window process to use `vscode-file` protocol and only disallow for OOPIF used for extension webview, then checking only `frame.processId` is sufficient.

bpasero commented on src/vs/platform/windows/electron-main/window.ts:
```diff
@@ -441,16 +443,45 @@ export class CodeWindow extends Disposable implements ICodeWindow {
+			}
+
+			if (details.frame?.url || details.frame?.parent) { // Coming from unrecognized frame
+				return false;
```
Should we log this condition too?

bpasero commented on src/vs/platform/windows/electron-main/window.ts:
```diff
@@ -441,16 +443,45 @@ export class CodeWindow extends Disposable implements ICodeWindow {
+
+			// Dealing with frame without a url set yet. This happens when first loading a window.
+			// Here we still want to allow loading the main workbench html files
+			if (details.url.startsWith(workbenchRootUrl) && uri.path.endsWith('.html')) {
```
Do we understand this properly?

deepak1556 replied:
I am also unsure which case does this represent ? Can you give an example of which frame is this referring to ?

mjbvz replied:
This is to handle the request to load the initial html contents of a window (typically this is `workbench.html`, but it can also be something like `processExplorer.html` )

Without this, this initial request ends up being blocked since `details.frame.url` is empty

bpasero reviewed (changes requested):
Scary change, I have to say I cannot really review this because I do not understand this Electron API well enough, Deepak should be the main driver.

mjbvz commented:
Thanks for the feedback. I just pushed an update. Unfortunately I had to move some code around which makes the change more difficult to review. If needed, I can split this change into separate PRs.

While trying to use @deepak1556's suggested approach, I ran into  bug: `CodeWindow` registers a `onBeforeRequestHandler` but can then be disposed of when the window is closed. When this happens, the handler remains active for other windows but any references to `this` inside of it now reference the disposed class. To fix this, I've moved all the request filtering logic into `app.ts` instead and register it on the `defaultSession` instead of on a web contents.

Then to fix the original issue, I added logic that checks the `routingId` of the request against the `routingId` of all known windows. Only windows with known routingIds are allowed to use `vscode-file`. In my testing, this blocks blocks `vscode-file` requests in subframes

deepak1556 commented on src/vs/code/electron-main/app.ts:
```diff
@@ -420,6 +422,91 @@ export class CodeApplication extends Disposable {
+			}
+
+			for (const window of this.windowsMainService.getWindows()) {
+				if (window.win && details.frame.routingId === window.win.webContents.mainFrame.routingId) {
```
`routingId` is only unique within a process, this handler will be triggered for requests from multiple processes. So you would also need to compare `processId` to make sure the frame matches.

Also, web worker extension host uses an iframe to host the contents and I think it uses `vscode-file` scheme for the top-level frame request. It will blocked by this handler since it is not the main frame in the window. We might either allow request to that web worker extension host initiator files or we can switch it to use a separate scheme (this will make it a cross-origin worker and will require https://github.com/microsoft/vscode/issues/87282#issuecomment-917539104 workaround)

mjbvz replied:
In my testing, the web worker iframe has the same processId as the main window. I believe this is because they are from the same origin?

Do you think this is a safe assumption? If not, I could also allow requests from iframes that are themselves served from `vscode-file` urls

bpasero commented on src/vs/code/electron-main/app.ts:
```diff
@@ -405,6 +405,8 @@ export class CodeApplication extends Disposable {
 		// Setup Auth Handler
 		this._register(appInstantiationService.createInstance(ProxyAuthHandler));
 
+		this.registerWebRequestFilter();
```
We already have a security related method `configureSession` above that we really call as first thing. Should we maybe move your new method into that one to have a single method for configuring security related things right from the beginning?

mjbvz commented:
@deepak1556 I've updated the code to check `details.frame.processId` against  `BrowserWindow.getAllWindows` This allows loading of the main VS Code windows, along with the shared process and the web worker. However it correctly blocks loading resources inside of webview

deepak1556 commented on src/vs/code/electron-main/app.ts:
```diff
@@ -153,6 +153,90 @@ export class CodeApplication extends Disposable {
+
+		const isAllowedVsCodeFileRequest = (details: Electron.OnBeforeRequestListenerDetails) => {
+			const frame = details.frame;
+			if (!frame || !this.windowsMainService) {
```
How does this condition accommodate resource requests from the web worker extension host ?

mjbvz replied:
In my testing, the web worker iframe has the same processId as the main window. I believe this is because they are from the same origin?

Do you think this is a safe assumption? If not, I could also allow requests from iframes that are themselves served from `vscode-file` urls

deepak1556 replied:
> In my testing, the web worker iframe has the same processId as the main window. I believe this is because they are from the same origin?

Oh okay then it should be fine.

deepak1556 reviewed (commented):
LGTM with one question
